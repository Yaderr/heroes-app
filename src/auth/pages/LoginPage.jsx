import { useNavigate } from "react-router-dom"

export const LoginPage = () => {
    
    const navigate = useNavigate()
    
    const onLogin = () => {
        navigate('/', {
            replace: true
        })
    }

    return (
        <>
            <h2 className="text-md">LoginPage</h2>
            <button onClick={ onLogin}>Login</button>
        </>
    )
}