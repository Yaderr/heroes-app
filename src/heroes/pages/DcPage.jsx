import { HeroList } from "../components/HeroList"

export const DcPage = () => {
    
    return (
        <>
            <h2 className="text-md font-bold">DcPage</h2>
            <br />

            <HeroList publisher="DC Comics" />
        </>
    )
}