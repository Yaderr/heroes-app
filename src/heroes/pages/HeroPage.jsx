import { Navigate, useNavigate } from 'react-router-dom'
import { useParams } from "react-router-dom"
import { getHeroById } from "../helpers"
import { useMemo } from 'react'

export const HeroPage = () => {

    const { id } = useParams()

    const hero = useMemo(() => getHeroById(id), [id])

    const navigator = useNavigate()

    const onNavigateBack = () => {
        navigator(-1)
    }

    if(!hero) {
        return (
            <Navigate to="/marvel" />
        )
    }
    
    return ( 
        <>
        <button className='animate__animated animate__fadeInLeft' onClick={onNavigateBack}>
            <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" strokeWidth={1.5} stroke="currentColor" className="w-6 h-6">
                <path strokeLinecap="round" strokeLinejoin="round" d="M10.5 19.5L3 12m0 0l7.5-7.5M3 12h18" />
            </svg>
        </button>
            <div className="flex justify-center animate__animated animate__fadeIn">
                <div>
                    <img className='rounded-lg border-2' src={`./heroes/${hero.id}.jpg`} alt="" />
                </div>
                <div className='ml-10 max-w-[500px]'>
                    <h1 className='text-2xl font-bold'>{ hero.superhero }</h1>
                    <ul className='space-y-2 mt-5 text-sm'>
                        <li><strong>Publisher:</strong> { hero.publisher } </li>
                        <li><strong>Alter ego:</strong> { hero.alter_ego } </li>
                        <li><strong>First appearance:</strong> { hero.first_appearance } </li>
                        <li className=''>
                            {
                                hero.characters.split(',').map(x => (<span key={x} className='bg-gray-200 mt-2 mr-2 p-2 rounded-md font-bold text-indigo-500 inline-block'>{x}</span>))    
                            }
                        </li>
                    </ul>
                </div>
            </div>
        </>
    )
}