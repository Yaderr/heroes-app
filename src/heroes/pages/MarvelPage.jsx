import { HeroList } from "../components/HeroList"

export const MarvelPage = () => {
    
    return (
        <>
            <h2 className="text-md font-bold">MarvelPage</h2>
            <br />

            <HeroList publisher="Marvel Comics" />
        </>
    )
}