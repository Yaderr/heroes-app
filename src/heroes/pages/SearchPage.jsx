import queryString from 'query-string'
import { useLocation, useNavigate } from "react-router-dom"
import { useForm } from "../../hooks/useForm"
import { HeroList } from "../components/HeroList"
import { favoritesReducer, searchHeroBySuperHero } from '../helpers'
import { HeroCard } from '../components/HeroCard'
import { useEffect, useReducer } from 'react'
import { types } from '../types/types'

const init = () => {
    return JSON.parse(localStorage.getItem('favorites') || '[]')
}

export const SearchPage = () => {

    const navigate = useNavigate()
    const location = useLocation()

    const { q='' } = queryString.parse(location.search)

    const { searchText, onInputChange} = useForm({
        searchText: q
    })

    const results = searchHeroBySuperHero(q)

    const onSearchSubmit = (event) => {
        event.preventDefault()

        if(searchText.trim().length <= 1) return 

        navigate(`?q=${searchText}`)        
    }

    const [favState, dispatch] = useReducer(favoritesReducer, [], init)

    const favIds = favState.map(fav => fav.id)

    useEffect(() => {
        localStorage.setItem('favorites', JSON.stringify(favState))
    },[favState])
    
    const onClickFav = (heroe, isFav) => {

        let action = {
            payload: heroe
        }

        if(isFav) {
            action.type = types.removeFavorites
        }else {
            action.type = types.addFavorite
        }
        
        dispatch(action)
    }
    
    return (
        <>
            <div className="flex justify-center">
               
                <form data-testid="search-form" className="flex bg-white drop-shadow-lg focus:drop-shadow-xl rounded-md py-2 px-2" onSubmit={onSearchSubmit}>
                    <input autoFocus name="searchText" onChange={onInputChange} value={ searchText } className="py-2 px-2 bg-transparent w-[500px] border-indigo-500 outline-none" placeholder="Buscar un heroe..." type="search" />
                    <button className="w-10 bg-indigo-500 text-white outline-none py-2 px-2 rounded-lg">
                        <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" strokeWidth={1.5} stroke="currentColor" className="w-6 h-6">
                            <path strokeLinecap="round" strokeLinejoin="round" d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z" />
                        </svg>
                    </button>
                </form>
            </div>
            
            <br />

            <div data-testid="default-message" style={{ display: q === ''? null: 'none'}} className="animate__animated animate__fadeIn flex h-[500px] w-full justify-center items-center">
                <span className='font-medium text-2xl text-indigo-500'>Busca un heroe...</span>
            </div>

            <div data-testid="no-heroes-found-message" style={{ display: results.length === 0 && q !== '' ? null : 'none' }} className="animate__animated animate__fadeIn flex h-[500px] w-full justify-center items-center">
                <span className='font-medium text-2xl text-indigo-500 text-center'>Oops <br />No hay heroes</span>
            </div>

            {
                results.map((hero) => (
                    <HeroCard key={hero.id} heroe={hero} isFav={favIds.includes(hero.id)} onClickFav={onClickFav} />
                ))
            }
        </>
    )
}