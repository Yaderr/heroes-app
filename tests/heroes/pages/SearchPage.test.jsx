import { fireEvent, render, screen } from "@testing-library/react"
import { SearchPage } from "../../../src/heroes/pages/SearchPage"
import { MemoryRouter, useNavigate } from "react-router-dom"

const mockUseNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate,
}))


describe('SerachPage Tests', () => {

    beforeEach(() => jest.clearAllMocks())
    
    test('debe mostrarse correctamente con los valores por defecto', () => {

        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot()
    })

    test('debe mostra a batman y en el input el valor del queryString', () => {
        
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('searchbox')
        expect(input.value).toBe('batman')

        const img = screen.getByTestId('heroe-img')
        expect(img.style.backgroundImage).toContain('batman')

        const defaultMessage = screen.getByTestId('default-message')
        expect(defaultMessage.style.display).toBe('none')

    })

    test('debe mostrar un error si no se encuentra el heroe' , () => {
        
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        )

        const notFoundHeroAlert = screen.getByTestId('no-heroes-found-message')

        expect(notFoundHeroAlert.style.display).toBeFalsy()
    })

    test('debe llamar el navigate a la pantalla nueva', () => {
        
        render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        )

        const inputSearch = screen.getByRole('searchbox')
        const form = screen.getByTestId('search-form')

        fireEvent.change(inputSearch, { target: { value: 'batman' }})
        fireEvent.submit(form)

        expect(mockUseNavigate).toHaveBeenCalledWith('?q=batman')
    })
})