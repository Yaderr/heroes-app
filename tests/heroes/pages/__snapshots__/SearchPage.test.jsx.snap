// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SerachPage Tests debe mostrarse correctamente con los valores por defecto 1`] = `
<div>
  <div
    class="flex justify-center"
  >
    <form
      class="flex bg-white drop-shadow-lg focus:drop-shadow-xl rounded-md py-2 px-2"
      data-testid="search-form"
    >
      <input
        class="py-2 px-2 bg-transparent w-[500px] border-indigo-500 outline-none"
        name="searchText"
        placeholder="Buscar un heroe..."
        type="search"
        value=""
      />
      <button
        class="w-10 bg-indigo-500 text-white outline-none py-2 px-2 rounded-lg"
      >
        <svg
          class="w-6 h-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </form>
  </div>
  <br />
  <div
    class="animate__animated animate__fadeIn flex h-[500px] w-full justify-center items-center"
    data-testid="default-message"
  >
    <span
      class="font-medium text-2xl text-indigo-500"
    >
      Busca un heroe...
    </span>
  </div>
  <div
    class="animate__animated animate__fadeIn flex h-[500px] w-full justify-center items-center"
    data-testid="no-heroes-found-message"
    style="display: none;"
  >
    <span
      class="font-medium text-2xl text-indigo-500 text-center"
    >
      Oops 
      <br />
      No hay heroes
    </span>
  </div>
</div>
`;
